### ISSUE TITLE
Versioning Strategy

### ISSUE THREAD

**Issue by shapeshed** [clearmatics/autonity#207]
As part of moving towards production we need to express a clear versioning strategy and ensure this is understood by Engineering, QA, Platform and Operational Teams.

The Engineering Team already widely use [Semantic Versioning][1] and it is proposed that this should be used for Autonity. The specification expresses clearly patch, minor and major releases and these map well onto release scenarios where patch and minor releases can be considered a soft fork and a major release is considered a hard fork. 

The relevant section from the specification follows

1. MAJOR version when you make incompatible API changes,
2. MINOR version when you add functionality in a backwards-compatible manner, and
3. PATCH version when you make backwards-compatible bug fixes.

To summarise

1. MAJOR - Hard Fork
2. MINOR - Soft Fork
3. PATCH - Soft Fork

This maps to

1. `1.x.x` -> `2.x.x` Hard Fork
2. `x.1.x` -> `x.2.x` Soft Fork
3. `x.x.1` -> `x.x.2` Soft Fork 

It is suggested that everyone reads and understands the SemVer scheme as this will form the basis for understanding which scenarios need to be used for testing and operating an Autonity release.

@flofrie @aiman Do you need any more discussion or documentation around this approach?

## References

* [Semantic Versioning Specification][1]

[1]: https://semver.org/

**Comment by flofrie** [comment]
@shapeshed  -- No further input or documentation from my end; I'm just really happy that we're moving to follow an explicit strategy! :-D

**Comment by yazzaoui** [comment]
Proposed slight modification :
- MAJOR version when you make incompatible API **or protocol** changes.

**Comment by shapeshed** [comment]
@yazzaoui agreed. have amended the original proposal - will move into the engineering internal site once @aiman has reviewed. he is on holiday today :beach_umbrella:

**Comment by rmsams** [comment]
Think the semver approach makes a ton of sense. How are we defining the *subject* of the version? As I see it, each of the following should have version numbers:

- Autonity System Class #1
- An Autonity System (eg "PlatformX", "Fnality Local System UK", etc)
- A sfw component (e.g, Transaction Broker, Asset Token Contract, etc.)

Is this everyone else's view too? If not, what is the policy?
